### PR TITLE
Shrink the navbar and hide the statusbar in landscape mode.

### PR DESCRIPTION
**Pull request by skunkmb**
To do so, create a new function `getRefreshedStyles` in `StylesProvider.js` for updating the height of the navbar. Whenever the orientation is updated, get the new styles from `getRefreshedStyles` and apply them.

I chose `40` as the landscape navbar height, but only for aesthetic reasons. If another value makes more sense, let me know!

(Done for the *"Issue #1187 Slimmer nav bar in landscape mode"* Google Code-in task.)

### CODE REVIEW

File: src/main/MainScreen.js [hunk 1]
```diff
@@ -22,12 +22,28 @@ export default class MainScreen extends PureComponent<Props> {
     this.props.navigation.navigate('DrawerOpen');
   };
 
+  constructor(props, context) {
+    super(props, context);
+    this.state = { shouldShowStatusBar: (
+      Dimensions.get('window').width < Dimensions.get('window').height
```
Comment (jainkuniya) [inline src/main/MainScreen.js]: you can get current orientation from app state.
like https://github.com/zulip/zulip-mobile/blob/master/src/account-info/AccountDetailsScreen.js#L48

File: src/main/MainScreen.js [hunk 2]
```diff
@@ -22,12 +22,28 @@ export default class MainScreen extends PureComponent<Props> {
...
     const { styles } = this.context;
 
     return (
-      <View style={[styles.flexed, styles.backgroundColor]}>
-        <ZulipStatusBar />
+      <View
+        onLayout={() => { this.reloadOrientation(); }}
+        style={[styles.flexed, styles.backgroundColor]}
+      >
+        { this.state.shouldShowStatusBar ? <ZulipStatusBar /> : null }
```
Comment (jainkuniya) [inline src/main/MainScreen.js]: just check here from `orientation`

File: src/nav/MainNavBar.js [hunk 3]
```diff
@@ -45,7 +54,10 @@ class MainNavBar extends PureComponent<Props> {
     const leftPress = editMessage ? actions.cancelEditMessage : onPressStreams;
 
     return (
-      <View style={[styles.navBar, { backgroundColor }]}>
+      <View
+        onLayout={() => { this.reloadOrientation(); }}
+        style={[this.state.navBarStyle, { backgroundColor }]}
```
Comment (jainkuniya) [inline src/nav/MainNavBar.js]: with the help of `orientation` conditionally apply height style.
like `<View style={[styles.navBar, { backgroundColor, height: orientation === 'portrait' ? PORTAIT_HEIGHT : LANDSCAPE_HEIGHT  }]}>`

File: src/main/MainScreen.js (hunk 2 above)
Comment (jainkuniya) [reply src/main/MainScreen.js]: And we can further simply it to `{orientation === 'portrait' && <ZulipStatusBar />}` (no `null` required)
like https://github.com/zulip/zulip-mobile/blob/master/src/chat/Chat.js#L67

**Comment by borisyankov** [comment]
Good job indeed.
You can split this in two. First, just hide the status bar. Then on a second consider the nav bar height (which might be ok as is though)

Do not forget to consider the safe area offset - which is to make it work on iPhone X (and possibly more devices later).

**Comment by skunkmb** [comment]
@jainkuniya Thanks for the suggestions! I redid almost everything in the new commits, ef38bc9 and a0a9ee5, with your ideas. I agree that it makes it a lot simpler!

**Comment by skunkmb** [comment]
@borisyankov Thanks! I tested it on iPhone X, and vertically it works fine:



Horizontally, the navbar is OK, but the notch cuts off part of the messages on the side (when taking screenshots in Xcode, for some reason it acts as if the notch isn't there, when really it cut off the `AM`s and the `in` in Steve Howell's message):



However, I also ran the simulator on the version of the app before my commits, and the notch-on-the-side problem was also apparent. 

Since it doesn't come as a consequence of the shrinking navbar, I was thinking that the notch side inset may be a bit out of the scope of this pull request; I could be wrong, though. Please let me know if I should work on it in this PR. I would need some guidance as to how to use the `safe area offset`. Thanks!

**Comment by skunkmb** [comment]
Thank you!

**Comment by skunkmb** [comment]
@borisyankov is this ready to be merged, or is there more I should work on?

File: src/main/MainScreen.js [hunk 4]
```diff
@@ -24,13 +26,18 @@ export default class MainScreen extends PureComponent<Props> {
 
   render() {
     const { styles } = this.context;
+    const { orientation } = this.props;
 
     return (
       <View style={[styles.flexed, styles.backgroundColor]}>
-        <ZulipStatusBar />
+        { orientation === 'PORTRAIT' && <ZulipStatusBar /> }
```
Comment (borisyankov) [inline src/main/MainScreen.js]: You should handle this from inside the `ZulipStatusBar` component.

File: src/main/MainScreen.js [hunk 5]
```diff
@@ -24,13 +26,18 @@ export default class MainScreen extends PureComponent<Props> {
...
         <MainNavBar onPressStreams={this.handlePressStreams} />
         <ChatContainer />
       </View>
     );
   }
 }
+
+export default connectWithActionsPreserveOnBack(state => ({
+  orientation: state.app.orientation,
```
Comment (borisyankov) [inline src/main/MainScreen.js]: ... and then this will be in the ZulipStatusBar component.

File: src/boot/StylesProvider.js [hunk 6]
```diff
@@ -36,7 +38,11 @@ class StyleProvider extends PureComponent<Props> {
   getChildContext() {
     const { theme } = this.props;
     const styles = StyleSheet.create(themeCreator(themeNameToObject[theme]));
-    return { theme, styles };
+    const navBarSizes = {
```
Comment (borisyankov) [inline src/boot/StylesProvider.js]: You don't need to use context to pass down these properties. (they are constant!)

File: src/nav/MainNavBar.js [hunk 7]
```diff
@@ -22,30 +22,34 @@ type Props = {
...
       editMessage,
+      orientation,
     } = this.props;
 
     const leftPress = editMessage ? actions.cancelEditMessage : onPressStreams;
+    const navBarSize = orientation === 'PORTRAIT' ? navBarSizes.portrait : navBarSizes.landscape;
 
     return (
-      <View style={[styles.navBar, { backgroundColor }]}>
+      <View style={[styles.navBar, { backgroundColor }, { height: navBarSize }]}>
```
Comment (borisyankov) [inline src/nav/MainNavBar.js]: You can have a navBarLandscape style that has the height set.

File: src/main/MainScreen.js (hunk 4 above)
Comment (skunkmb) [reply src/main/MainScreen.js]: @borisyankov OK, I moved it inside the component in 5adbfce. Thanks!

File: src/nav/MainNavBar.js (hunk 7 above)
Comment (skunkmb) [reply src/nav/MainNavBar.js]: @borisyankov OK, I created the `navBarLandscape` style in 0d86fe4. Thanks!

File: src/boot/StylesProvider.js (hunk 6 above)
Comment (skunkmb) [reply src/boot/StylesProvider.js]: @borisyankov OK, now they are in their own style in 0d86fe4, like you suggested, so they are not in `context` anymore. Thanks!

**Comment by skunkmb** [comment]
@borisyankov OK, the tests are passing now. I also split this into 2 PRs like you suggested, #1658 and #1659. Tests are passing there as well, too. 

Please let me know if there is more I should work on here, or if I should close this PR since it is now redundant. Thanks!

**Comment by borisyankov** [comment]
The PR still feels like too much change, too much work and too much chance of regression for not enough benefit. We can reconsider this, but closing for now.

**Pull request closed without merging**